src/sort/fn.rs

```diff
@@ -0,0 +1,347 @@
+        value: Value,
+        extractor: &Extractor,
+        termdag: &mut TermDag,
+    ) -> Option<(Cost, Expr)> {
+        let ValueFunction(name, inputs) = ValueFunction::load(self, &value);
+        let (cost, args) = inputs.into_iter().try_fold(
+            (1usize, vec![Expr::Lit((), Literal::String(name))]),
+            |(cost, mut args), (sort, value)| {
+                let (new_cost, term) = extractor.find_best(value, termdag, &sort)?;
+                args.push(termdag.term_to_expr(&term));
+                Some((cost.saturating_add(new_cost), args))
+            },
+        )?;
+
+        Some((cost, Expr::call("unstable-fn", args)))
+    }
+}
+
+impl IntoSort for ValueFunction {
+    type Sort = FunctionSort;
+    fn store(self, sort: &Self::Sort) -> Option<Value> {
+        let mut functions = sort.functions.lock().unwrap();
+        let (i, _) = functions.insert_full(self);
+        Some(Value {
+            tag: sort.name,
+            bits: i as u64,
+        })
+    }
+}
+
+impl FromSort for ValueFunction {
+    type Sort = FunctionSort;
+    fn load(sort: &Self::Sort, value: &Value) -> Self {
+        sort.get_value(value)
+    }
+}
+
+/// Takes a string and any number of partially applied args of any sort and returns a function
+struct FunctionCTorTypeConstraint {
+    name: Symbol,
+    function: Arc<FunctionSort>,
+    string: Arc<StringSort>,
+}
+
+impl TypeConstraint for FunctionCTorTypeConstraint {
+    fn get(&self, arguments: &[AtomTerm]) -> Vec<Constraint<AtomTerm, ArcSort>> {
+        // Must have at least one arg (plus the return value)
+        if arguments.len() < 2 {
+            vec![Constraint::Impossible(
+                constraint::ImpossibleConstraint::ArityMismatch {
+                    atom: core::Atom {
+                        head: self.name,
+                        args: arguments.to_vec(),
+                    },
+                    expected: 2,
+                    actual: arguments.len(),
+                },
+            )]
+        } else {
+            vec![
+                Constraint::Assign(arguments[0].clone(), self.string.clone()),
+                Constraint::Assign(
+                    arguments[arguments.len() - 1].clone(),
+                    self.function.clone(),
+                ),
+            ]
+        }
+    }
+}
+
+// (unstable-fn "name" [<arg1>, <arg2>, ...])
+struct Ctor {
+    name: Symbol,
+    function: Arc<FunctionSort>,
+    string: Arc<StringSort>,
+}
+
+impl PrimitiveLike for Ctor {
+    fn name(&self) -> Symbol {
+        self.name
+    }
+
+    fn get_type_constraints(&self) -> Box<dyn TypeConstraint> {
+        Box::new(FunctionCTorTypeConstraint {
+            name: self.name,
+            function: self.function.clone(),
+            string: self.string.clone(),
+        })
+    }
+
+    fn apply(&self, values: &[Value], egraph: Option<&mut EGraph>) -> Option<Value> {
+        let egraph = egraph.expect("Cannot create function in fact only in actions");
```

Comment:
This error message should be improved
Ideally, we should check that we don't use this primitive in queries at all at typechecking time

Reply:
I rewrote it a bit, let me know if you have any suggestions on how to make it clearer.

I think, although correct me if I am wrong, that the easiest way to make this fail during type-checking would be to modify [`Facts.to_query`](https://github.com/egraphs-good/egglog/blob/83a51a9f6f1ad69b5bd067558dd7e7a29cbefbbf/src/ast/mod.rs#L1000-L1007) and [`GenericExpr.to_query`](https://github.com/egraphs-good/egglog/blob/83a51a9f6f1ad69b5bd067558dd7e7a29cbefbbf/src/core.rs#L516-L523) to return `Result`s so that we can return a type-checking error if you try to create a query with these functions.

I am hesitant to make that larger change for just this special case, since it's still unstable.